NSWorkspace openFile doesn't support .app and directories properly

onflapp opened the pull request:
NSWorkspace openFile should be able to handle different types of paths.
- if path is an app, launch it
- if path is a bundle, open it in associated app
- it path is regular folder, open it in file view

The current implementation of NSWorkspace openFile does not handle all cases in consistent way.
openFile:fromImage:at:inView seems to be much more complete.
This pull request combines both methods in one.

trunkmaster commented:
@onflapp could you please explain what exact cases is not handled by Workspace in wrong way? Plus I can't find what your changes are because you've moved around big portions of existing code. Could you please make changes to existing code without 2 extra private methods or explain what exactly are you trying to fix?

onflapp commented:
The problem is that [NSWorkspace openFile:withApplication:andDeactivate] does not work with directories and apps properly. Current implementation does not check whether path is a directory, for example ("gopen -o $PWD" opens directory in TextEdit).

Related method [NSWorkspace openFile:fromImage:at:inView] works as expected.

In order to fix this problem, one option is to copy and adapt the implementation from openFile:fromImage:at:inView, however this will lead to code duplication.

The proposed solution is to declare private "super" method [NSWorkspace _doOpenFile:fromImage:at:inView:withApp:deactivate] which combines functionality needed by all public methods. As the super method is rather large, I implemented part of shared functionality in [NSWorkspace _connectToApplication:openFile:andDeactivate].

trunkmaster commented:
OK, I think some points needs clarification:
1. `gopen` uses GNUstep GUI NSWorkspace method `-openFile:withApplication` and has no connection to Controller+NSWorkspace method. I've started Controller+NSWorkspace as a delegate for NSWorkspace (GSWorkspaceApplication preference). Although it is not finished yet. Actually at this point in time Controller+NSWorkspace is a separate implementation of "NSWorkspace" parrallel to GNUstep's. By saing that, I think this PR doesn't fix `gopen` usage and "GSWorkspaceApplication" setting change is premature.
2. You want to offer improvement to Controller+NSWorkspace methods. I have some notes:
   - Contrlroller already has `-_connectApplication` method. Why we need another one?
   - there are 4 methods related to `-openFile` action. Two of them (`-openFile:` and `-openFile:withApplication:`) just call `-openFile:fromImage:at:inView:`. These methods behaviour is correct.
   - `-openFile:withApplication:andDeactivate:` seems to have problem - if `appName` was not set it calls "TextEdit" application. I guess this is what you're trying to fix. But, keep in mind, plaintext files always should be opened in TextEdit if no associated application for this file type.

Some time ago I've modified OpenUp to open uncompressed archive directory in Workspace. Look at this code in ApplicationDelegate+decompression.m:
```objective-c
[[NSWorkspace sharedWorkspace] openFile:unarchiveDirectoryPath 
                        withApplication:@"Workspace"];
```
It is temporary solution, but it works for now.

onflapp commented:
@trunkmaster I know what you are saying. However, my assumption and my expectation is that if I call [NSWorkspace openFile:] (and all it's different variations) it does exactly what Workspace app would do. I would not treat them separately under NEXTSPACE. 

In GNUStep, yes, this makes sense because you want to allow for different implementations but in context of NEXTSPACE my expectation is that NSWorkspace is the Workspace app, as it would be in Next/OpenStep or Mac.

trunkmaster commented:
@onflapp I've played a bit around [NSWorkspace openFile:] and found that this method call goes down to [Controller openFile:withApplication:andDeactivate] method if `GSWorkspaceApplication == Workspace`. I think fix should be applied to `if (appName == nil) {}` block inside Controller's method instead of introducing 2 extra private methods (one of which already exists as `-_connectApplication`).

onflapp commented:
Please take a look at openFile:fromImage:at:inView: as well.
In addition to what openFile:withApplication:andDeactivate is doing, it does check for path being an app or directory.

It would be possible to copy all that to openFile:withApplication:andDeactivate  but then you just end up with duplicating non-trivial code in two places.

trunkmaster commented:
I think the same result can be achieved in another way without tossing bunch of code with no noticable changes inside that code. Try to replace code in `-openFile:withApplication:andDeactivate:` starting from line 445:
```objective-c
    if ([self _extension:[fullPath pathExtension] role:nil app:&appName] == NO) {
      appName = [[NXTDefaults userDefaults] objectForKey:@"DefaultEditor"];
      if (!appName || [appName isEqualToString:@""]) {
        appName = @"TextEdit";
      }
    }
```
with
```objective-c
    [self openFile:fullPath fromImage:nil at:NSZeroPoint inView:nil];
```

This is what you're trying to get, isn't it?

onflapp commented:
well, almost :-)
that was actually my first try.

but what happens if you need to feed in the "appName" or "deactivate"?

that's why ended up with openFile:fromImage:at:inView:withApplication:deactivate as private method.

My implementation is a bit ugly, I agree, but what I was trying to do is to have one method making all decisions which can be called from different entry points. Calling [NSWorkspace openFIle:] should not really be different from clicking icon in the dock or using service. It should make the same decision.

trunkmaster commented:
> well, almost :-)
> that was actually my first try.

No problem. I want to point you to several problems with this PR:
1. Your commit hides actual changes in code logic (because of massive code removal and addition).
2. You've added extra methods which are not needed actually.
3. You've added method that duplicates existing one (-connectApplication:).

> but what happens if you need to feed in the "appName" or "deactivate"?

You should call `-openFile:withApplication:andDeactivate:` directly. Here we are talking about [NSWorkspace openFile:] call and everything works as expected.

> that's why ended up with openFile:fromImage:at:inView:withApplication:deactivate as private method.
> 
> My implementation is a bit ugly, I agree, but what I was trying to do is to have one method making all decisions which can be called from different entry points. Calling [NSWorkspace openFIle:] should not really be different from clicking icon in the dock or using service. It should make the same decision.

It is with my proposed change.

onflapp commented:
Unfortunately, this does not solve all problems. Let me restate the issues 
The following uses of NSWorkspace do not work correctly:

`[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents"];`
-> should open/reopen directory viewer in Workspace.app

`[[NSWorkspace sharedWorkspace] openFile:@"/Application/Librarian.app"];`
-> should launch application Librarian

`[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents" withApplication:@"Librarian"];`
-> should open directory in application Librarian and make it active

`[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents" withApplication:@"Librarian" andDeactivate:YES];`
-> should open directory in application Librarian, but not make it active

trunkmaster commented:
> `[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents" withApplication:@"Librarian" andDeactivate:YES];`
> -> should open directory in application Librarian, but not make it active

Wrong. `andDeactivate:` corresponds to application that make this call.

Anyway I've created locally quick and dirty tests - everything works as expected with my proposed change. Check it out.

onflapp commented:
@trunkmaster unfortunately following cases do not work as expected
`[[NSWorkspace sharedWorkspace] openFile:@"/Application/Librarian.app"]`
`[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents"]`

onflapp commented:
@trunkmaster  I cleanup the code changes so that it is easier for you to follow:

1. I added one single method that does all necessary checks based on parameters you pass to it
**`openFile:fromImage:at:inView:withApplication:andDeactivate:`**
and takes whatever action makes sense

2. moved implementation code from
` openFile:withApplication:andDeactivate:` 
to
`connectionApplication:openFile:andDeactivate:`
which is then called by the new method **`openFile:fromImage:at:inView:withApplication:andDeactivate:`**

3. changed methods
` openFile:withApplication:andDeactivate:` 
and
`openFile:fromImage:at:inView:`

to simply call the new method **`openFile:fromImage:at:inView:withApplication:andDeactivate:`**
these methods should not do any decision/action on its own


In addition
1. I changed the **System/Library/Preferences/GlobalDefaults.plist** to make Workspace app the default Workspace service.

2. modified the **openFileViewerService** handler to not make any decision on its own, but call **`openFile:`**, which in tern calls  **`openFile:fromImage:at:inView:withApplication:andDeactivate:`**

trunkmaster commented:
@onflapp do you apply my proposed change? Please make tests with "nsworkspace" branch and report your results.
> 
> @trunkmaster unfortunately following cases do not work as expected
> `[[NSWorkspace sharedWorkspace] openFile:@"/Application/Librarian.app"]`

I don't have Librarian.app (do you?) so I used Review.app. I've got started Review.app. What is your result? What did you expect?

> `[[NSWorkspace sharedWorkspace] openFile:@"/home/Documents"]`

I've got new File Viewer with specified path. What is your result?

onflapp commented:
@trunkmaster, this is strange.

OK, I'll re-test. Sorry about this.

onflapp commented:
@trunkmaster, I can confirm your change works. Many thanks!
I forgot to change the GSWorkspaceApplication settings.

I agree, this is cleaner approach.
I changed the pull request to use your implementation.
